lib/class-wp-style-engine.php

```diff
@@ -0,0 +1,69 @@
+<?php
+/**
+ * WP_Style_Engine class
+ *
+ * @package Gutenberg
+
+ */
+
+/**
+ * Singleton class representing the style engine.
+ *
+ * Consolidates rendering block styles to reduce duplication and streamline
+ * CSS styles generation.
+ *
+ * @since 6.0.0
+ */
+class WP_Style_Engine_Gutenberg {
+	/**
+	 * Registered CSS styles.
+	 *
+	 * @since 5.5.0
+	 * @var array
+	 */
+	private $registered_styles = array();
+
+	/**
+	 * Container for the main instance of the class.
+	 *
+	 * @since 5.5.0
+	 * @var WP_Style_Engine_Gutenberg|null
+	 */
+	private static $instance = null;
+
+	public function __construct() {
+		// Borrows the logic from `gutenberg_enqueue_block_support_styles`.
+		$action_hook_name = 'wp_footer';
+		if ( wp_is_block_theme() ) {
+			$action_hook_name = 'wp_enqueue_scripts';
+		}
+		add_action(
+			$action_hook_name,
+			array( $this, 'output_styles' )
+		);
+	}
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	public function add_style( $key, $value ) {
+		$this->registered_styles[ $key ] = $value;
```

Comment:
Thanks for getting this started!

Just looking at deduping/storying common values, such as the following, and wondering if we can do it as we add them or whether it can be done on output.

There are quite a few of them generated by layout:

```css
.wp-container-14 > :where(:not(.alignleft):not(.alignright)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-container-14 > .alignwide { max-width: 1000px;}.wp-container-14 .alignfull { max-width: none; }.wp-container-14 .alignleft { float: left; margin-right: 2em; margin-left: 0; }.wp-container-14 .alignright { float: right; margin-left: 2em; margin-right: 0; }.wp-container-14 > * { margin-top: 0; margin-bottom: 0; }.wp-container-14 > * + * { margin-top: var( --wp--style--block-gap );  margin-bottom: 0; }



.wp-container-15 > :where(:not(.alignleft):not(.alignright)) {max-width: 650px;margin-left: auto !important;margin-right: auto !important;}.wp-container-15 > .alignwide { max-width: 1000px;}.wp-container-15 .alignfull { max-width: none; }.wp-container-15 .alignleft { float: left; margin-right: 2em; margin-left: 0; }.wp-container-15 .alignright { float: right; margin-left: 2em; margin-right: 0; }.wp-container-15 > * { margin-top: 0; margin-bottom: 0; }.wp-container-15 > * + * { margin-top: var( --wp--style--block-gap );  margin-bottom: 0; }
```

I might have a play around if that's okay.

Reply:
Thanks for taking a look @ramonjd! It's a very small step in digging in and seeing what we might be able to do 😀  

> and wondering if we can do it as we add them or whether it can be done on output.

Yes, that'd be a great next step, I haven't _quite_ started looking at that kind of de-duping yet, but feel free to jump in, the more the merrier! I was wondering if we can construct more semantic-like (or at least non-random) classnames, then perhaps using the class name as a key might be a way to naturally de-dupe... I'm sure there could be lots of issues with attempting that, but might be one way to explore it 🙂

Reply:
>  I was wondering if we can construct more semantic-like (or at least non-random) classnames, then perhaps using the class name as a key might be a way to naturally de-dupe... I'm sure there could be lots of issues with attempting that, but might be one way to explore 

For sure! 

I'm starting with exploring different registry models, which might make it easier to parse/dedupe registered styles, e.g.,

```php
			WP_Style_Engine_Gutenberg::get_instance()->add_styles(
				array(
					"$selector > :where(:not(.alignleft):not(.alignright))" =>
					array(
						'max-width'    => esc_html( $all_max_width_value ),
						'margin-left'  => 'auto !important',
						'margin-right' => 'auto !important',
					),
				)
			);
```

Not sure where it's going yet, but starting like that 😄

Reply:
Oh, nice! The array based approach looks pretty cool. I started having a play with returning the class name from `add_style` so that we can start building out an array of class names to add, so thought I might try something like the following:

```php
			$class_names[] = $style_engine->add_style(
				"wp-container-all-max-width-$all_max_width_value",
				".wp-container-all-max-width-$all_max_width_value > :where(:not(.alignleft):not(.alignright)) {" .
				'max-width: ' . esc_html( $all_max_width_value ) . ';' .
				'margin-left: auto !important;' .
				'margin-right: auto !important;' .
				'}'
			);
```

It was starting to look pretty messy, though! I quite like your idea of an array to split it out to make it much more readable. Looking forward to comparing notes!

Reply:
Inspired by your array idea @ramonjd, I've had a play at writing an `add_style` function that looks a bit like the following:

```php
	public function add_style( $key, $options ) {
		$class    = ! empty( $options['suffix'] ) ? $key . '-' . sanitize_title( $options['suffix'] ) : $key;
		$selector = ! empty( $options['selector'] ) ? ' ' . trim( $options['selector'] ) : '';
		$rules    = ! empty( $options['rules'] ) ? $options['rules'] : array();
		$prefix   = ! empty( $options['prefix'] ) ? $options['prefix'] : '.';

		if ( ! $class ) {
			return;
		}

		$style = "{$prefix}{$class}{$selector} {\n";

		if ( is_string( $rules ) ) {
			$style .= '  ';
			$style .= $rules;
		} else {
			foreach( $rules as $rule => $value ) {
				$style .= "  {$rule}: {$value};\n";
			}
		}
		$style .= "}\n";

		$this->registered_styles[ $class . $selector ] = $style;

		return $class;
	}
```

This then allows us to call it like:

```php
			// Add value specific content size.
			$class_names[] = $style_engine->add_style(
				'wp-layout-default-content-size',
				array(
					'suffix'   => $all_max_width_value,
					'selector' => '> :where(:not(.alignleft):not(.alignright))',
					'rules' => array(
						'max-width'    => esc_html( $all_max_width_value ),
						'margin-left'  => 'auto !important',
						'margin-right' => 'auto !important',
					)
				)
			);
```

To then output something like the following:

```CSS
.wp-layout-default-content-size-650px > :where(:not(.alignleft):not(.alignright)) {
  max-width: 650px;
  margin-left: auto !important;
  margin-right: auto !important;
}
```

In view-source, for the default layout, in my local environment it's currently looking like this:

![image](https://user-images.githubusercontent.com/14988353/155649535-1ade7d4c-66e8-49d4-a69a-0de45ad9b3a5.png)

I'll continue hacking around with `layout.php` to see how this might look (it could get a bit more complex with the `flex` layout), but I just thought I'd share where I'm up to in case I run out of time before the end of the week 😀

I _think_ by combining the class name + a slugified unique value + selector for the key for the style, we should be able to de-dupe the styles somewhat reliably in the style array?

Reply:
Thanks @andrewserong 

Very interesting! I think including the extra metadata in options is an excellent idea for extensibility/readability.

I'm wondering if we should store the styles in `array` format up until the time we want to use them in case we need (or allow via hooks) any preprocessing. I suppose this sort of stuff will become clear later.

I'm also getting similar results doing the following:

```php
	public function add_styles( $styles ) {
		if ( ! is_array( $styles ) ) {
			return array();
		}

		foreach ( $styles as $key => $value ) {
			$value = is_array( $value ) ? $value : array();
			if ( isset( $this->registered_styles[ $key ] ) ) {
				$this->registered_styles[ $key ] = array_merge( $this->registered_styles[ $key ], $value );
			} else {
				$this->registered_styles[ $key ] = $value;
			}
		}
	}

	public function output_styles() {
		$callback = function( $css_selector, $css_ruleset ) {
			$style = "{$css_selector} {\n";
			foreach ( $css_ruleset as $css_property => $css_value ) {
				$style .= "    {$css_property}: {$css_value};\n";
			}
			$style .= "}\n";
			return $style;
		};

		$output = array_map( $callback, array_keys( $this->registered_styles ), array_values( $this->registered_styles ) );
		$output = implode( "\n", $output );

		echo "<style>\n$output</style>\n";
	}
```

Adding styles:

```php

$styles = array();
$styles[ "$selector > :where(:not(.alignleft):not(.alignright))" ] = array(
    'max-width'    => esc_html( $all_max_width_value ),
    'margin-left'  => 'auto !important',
    'margin-right' => 'auto !important',
);
WP_Style_Engine_Gutenberg::get_instance()->add_styles( $styles );
```

Output example

```css
.wp-container-5 > :where(:not(.alignleft):not(.alignright)) {
    max-width: 650px;
    margin-left: auto !important;
    margin-right: auto !important;
}
```

Just got bogged down trying to dedupe, so there's probably a better model or pattern.

Reply:
Ah, nice! Yes, it could be worth keeping the array around so that the method of output can be customised (or possibly filtered, even?) — It'll be interesting to explore further.

I've pushed my exploration in https://github.com/WordPress/gutenberg/pull/38974/commits/c1b24c222558b7c368eb988bda519043f25969d9 which also removes the randomly generated class name, but the trade-off there is that we then have more verbose class names added to the markup, which may or may not be a good or a bad thing 😄🤷 

<img width="1231" alt="image" src="https://user-images.githubusercontent.com/14988353/155657253-030aefb7-ec58-4d15-9663-0fc1ab69bd1f.png">

Reply:
Of course, the good thing about removing the randomly generated class name is that it's then fairly trivial to `de-dupe` as we never wind up with duplicates in the `registered_styles` array.

Reply:
> removes the randomly generated class name, but the trade-off there is that we then have more verbose class names added to the markup

Nice!! 

I like the approach, but I think we'd have to be careful about the classnames we generate. Folks could start relying on `wp-layout-default-wide-size-1000px`! 

But if used in conjunction with standardized naming schemes it would be ace.